File: src/kernel/floors.js
/**
 * WEALTH by arifOS — F1–F13 Floor Enforcement Engine
 *
 * Physics > Narrative. Maruah > Convenience.
 * Every operation MUST pass through checkFloors() before execution.
 *
 * DITEMPA BUKAN DIBERI — 999 SEAL ALIVE
 */

export const FLOORS = {
  F1:  { id: 'F1',  name: 'Reversible',          rule: 'All mutations must be reversible or declare irreversibility' },
  F2:  { id: 'F2',  name: 'Truth Band',           rule: '≥99% truth or declare confidence band explicitly' },
  F3:  { id: 'F3',  name: 'Human-AI-Evidence',    rule: 'Human judgment, AI analysis, and evidence must align' },
  F4:  { id: 'F4',  name: 'Clarity',              rule: 'Every output reduces complexity ΔS≤0, never adds noise' },
  F5:  { id: 'F5',  name: 'Peace',                rule: 'Peace score ≥ 1.0 before seal. No unresolved panic.' },
  F6:  { id: 'F6',  name: 'Maruah-First',         rule: 'ASEAN/MY dignity context. Maruah is a real variable.' },
  F7:  { id: 'F7',  name: 'Humility Band',        rule: 'Uncertainty declared: 0.03–0.15 minimum on projections' },
  F8:  { id: 'F8',  name: 'Law & Safety',         rule: 'Financial data stays local. No external sync without consent.' },
  F9:  { id: 'F9',  name: 'Anti-Hantu',           rule: 'No phantom balances, ghost transactions, or unresolved entries' },
  F10: { id: 'F10', name: 'AI Ontology',          rule: 'AI advises only. Human decides. AI never claims authority.' },
  F11: { id: 'F11', name: 'Auth Critical',        rule: 'Critical commands require authentication + 888 HOLD' },
  F12: { id: 'F12', name: 'Block Overrides',      rule: 'No bypass of floor logic permitted, by anyone' },
  F13: { id: 'F13', name: 'Sovereign Human Veto', rule: 'Human can reject any AI recommendation, unconditionally' },
};

export const EPISTEMIC = {
  CLAIM:      'CLAIM',
  PLAUSIBLE:  'PLAUSIBLE',
  HYPOTHESIS: 'HYPOTHESIS',
  ESTIMATE:   'ESTIMATE',
  UNKNOWN:    'UNKNOWN',
};

export const HOLD = {
  TRIGGERED: '888-HOLD-TRIGGERED',
  CLEARED:   '888-HOLD-CLEARED',
  PENDING:   '888-HOLD-PENDING',
};

/**
 * checkFloors — Gate function. Call before ANY financial operation.
 * @param {Object} operation - { type, payload, epistemic, reversible }
 * @returns {{ pass: boolean, holds: string[], violations: string[], warnings: string[] }}
 */
export function checkFloors(operation) {
  const result = {
    pass: true,
    holds: [],
    violations: [],
    warnings: [],
    epoch: new Date().toISOString(),
  };

  // F1 — Reversible
  if (operation.reversible === false) {
    result.holds.push(HOLD.TRIGGERED);
    result.holds.push(`F1: Operation "${operation.type}" declared irreversible. 888 HOLD activated.`);
    result.pass = false;
  }

  // F2 — Truth Band
  if (operation.epistemic === EPISTEMIC.UNKNOWN) {
    result.warnings.push('F2: UNKNOWN epistemic tag. Declare confidence band before proceeding.');
  }
  if (operation.confidence !== undefined && operation.confidence < 0.99 && !operation.band_declared) {
    result.warnings.push(`F2: Confidence ${operation.confidence} < 0.99 — band must be declared.`);
  }

  // F5 — Peace
  if (operation.peace2 !== undefined && operation.peace2 < 1.0) {
    result.holds.push(`F5: Peace score ${operation.peace2} < 1.0. Resolve state before seal.`);
    result.pass = false;
  }

  // F7 — Humility Band on projections
  if (operation.type === 'PROJECTION' && !operation.uncertainty_band) {
    result.violations.push('F7: Projection without uncertainty band declared. Violation.');
    result.pass = false;
  }

  // F8 — No external sync without consent
  if (operation.external_sync === true && !operation.user_consent) {
    result.violations.push('F8: External sync attempted without user consent. BLOCKED.');
    result.pass = false;
  }

  // F9 — Anti-Hantu
  if (operation.has_unresolved_entries === true) {
    result.warnings.push('F9: Unresolved entries detected. Resolve before seal.');
  }

  // F10 — AI Ontology
  if (operation.ai_is_deciding === true) {
    result.violations.push('F10: AI attempting to decide, not advise. BLOCKED.');
    result.pass = false;
  }

  // F12 — Block overrides
  if (operation.floor_override === true) {
    result.violations.push('F12: Floor override attempted. BLOCKED. No exceptions.');
    result.pass = false;
  }

  return result;
}

/**
 * assertFloors — Hard assert. Throws if floors fail.
 * @param {Object} operation
 */
export function assertFloors(operation) {
  const result = checkFloors(operation);
  if (!result.pass) {
    throw new Error(
      `FLOOR VIOLATION\n` +
      `Violations: ${result.violations.join(' | ')}\n` +
      `Holds: ${result.holds.join(' | ')}\n` +
      `Epoch: ${result.epoch}`
    );
  }
  return result;
}

File: host/governance/floors.py
"""
F1–F13 Constitutional Floor Evaluation
Ported from embedded JS theory (host/kernel/floors.js, canon/GOVERNANCE.md)
"""

from typing import Any, Dict, List, Optional
from datetime import datetime

FLOORS = {
    "F1": "Amanah — reversible or explicitly irreversible",
    "F2": "Truth — epistemic tag required",
    "F3": "Input Clarity — clear task definition",
    "F4": "Clarity — reduce complexity",
    "F5": "Peace — no unresolved panic",
    "F6": "Maruah — dignity as variable",
    "F7": "Humility — uncertainty band",
    "F8": "Law — local-first data",
    "F9": "Anti-Hantu — no phantom entries",
    "F10": "AI Only Advises — human decides",
    "F11": "Auth — PIN for critical",
    "F12": "No Override — floors unbypassable",
    "F13": "Human Veto — final authority",
}


def check_floors(args: Dict[str, Any]) -> Dict[str, Any]:
    """
    Evaluate F1–F13 constitutional floors.
    Returns: pass, verdict, violations, holds, warnings, epistemic, vault_log_entry, witness
    """
    violations: List[str] = []
    holds: List[str] = []
    warnings: List[str] = []

    reversible = args.get("reversible")
    human_confirmed = args.get("human_confirmed", False)
    epistemic = args.get("epistemic", "ESTIMATE")
    ai_is_deciding = args.get("ai_is_deciding", False)
    floor_override = args.get("floor_override", False)
    peace2 = args.get("peace2", 1.0)
    maruah_score = args.get("maruah_score", 0.5)
    uncertainty_band = args.get("uncertainty_band")
    operation_type = args.get("operation_type", "PROJECTION")
    scale_mode = args.get("scale_mode", "enterprise")

    # F1 — Amanah
    if reversible is False and not human_confirmed:
        holds.append("F1: 888_HOLD — Irreversible action requires human confirmation.")

    # F2 — Truth
    if epistemic == "UNKNOWN":
        warnings.append("F2: Epistemic state is UNKNOWN.")
    confidence = args.get("confidence")
    if confidence is not None and confidence < 0.99 and uncertainty_band is None:
        warnings.append("F2: Confidence < 0.99 and no uncertainty band declared.")

    # F3 — Input Clarity
    task_definition = args.get("task_definition")
    if task_definition is not None and len(str(task_definition).strip()) < 10:
        warnings.append("F3: Input lacks clarity (task definition too short).")

    # F5 — Peace
    if peace2 < 1.0:
        holds.append("F5: Peace² below 1.0.")

    # F6 — Maruah
    maruah_floor = args.get("maruah_floor", 0.6)
    if maruah_score < maruah_floor:
        holds.append(f"F6: Maruah score {maruah_score} below floor {maruah_floor}.")

    # F7 — Humility
    if operation_type == "PROJECTION" and uncertainty_band is None:
        warnings.append("F7: Projection missing uncertainty band.")

    # F10 — AI Ontology
    if ai_is_deciding:
        violations.append("F10: AI cannot be the final decision maker.")

    # F12 — No Override
    if floor_override:
        violations.append("F12: Floor override is not permitted.")

    # F13 — Human Veto (escalated for high-scale irreversible operations)
    if not human_confirmed and reversible is False and scale_mode in {"national", "crisis", "civilization"}:
        holds.append("F13: High-scale irreversible action requires human confirmation.")

    # F8 / F9 / F11 — placeholder checks using explicit flags
    if args.get("phantom_entries"):
        violations.append("F9: Phantom entries detected.")
    if args.get("critical") and not args.get("pin_verified"):
        holds.append("F11: PIN required for critical operation.")

    # Derive verdict
    if violations:
        verdict = "VOID"
    elif holds:
        verdict = "HOLD"
    elif warnings:
        verdict = "CAUTION"
    else:
        verdict = "SEAL"

    return {
        "pass": verdict in ("SEAL", "CAUTION"),
        "verdict": verdict,
        "violations": violations,
        "holds": holds,
        "warnings": warnings,
        "epistemic": epistemic,
        "vault_log_entry": {"tool": "wealth_check_floors", "epoch": datetime.utcnow().isoformat() + "Z"},
        "witness": {"human": human_confirmed, "ai": True, "earth": True},
    }


def maruah_band(maruah_score: float) -> str:
    if maruah_score >= 0.85:
        return "SOVEREIGN"
    if maruah_score >= 0.70:
        return "STABLE"
    if maruah_score >= 0.60:
        return "FLOOR"
    if maruah_score >= 0.40:
        return "AMBER"
    return "RED"
